### ISSUE TITLE
[TypeScript]: Syntax error on "export default interface" syntax

### ISSUE THREAD

**Issue by alangpierce** [babel/babel#7118]
Choose one: is this a bug report or feature request?
**Bug report.**

### Input Code
```js
export default interface A {}
```

[Parse error in Babel repl](http://babeljs.io/repl/#?babili=false&browsers=&build=&builtIns=false&code_lz=KYDwDg9gTgLgBAE2AMwIYFcA28CWA7GYKNAY2DgEE4BvAXyA&debug=false&circleciRepo=&evaluate=true&lineWrap=false&presets=es2015%2Ces2016%2Ces2017%2Creact%2Cstage-0%2Cstage-1%2Cstage-2%2Cstage-3%2Ces2015-loose%2Ctypescript&prettier=true&targets=Node-8&version=7.0.0-beta.36)
[Works in TypeScript repl](https://www.typescriptlang.org/play/index.html#src=export%20default%20interface%20A%20%7B%7D)

### Babel/Babylon Configuration (.babelrc, package.json, cli command)
7.0.0-beta.35 repl with TypeScript transform enabled.

### Expected Behavior
The syntax was added in https://github.com/Microsoft/TypeScript/pull/16040, so Babel should accept it (and no code should be emitted since it's a type export).

### Current Behavior
Babylon gives a parse error saying it's an unexpected token.

### Possible Solution
Probably something similar to #7075 😄

### Context
I have a TypeScript project that I'm trying to run through a custom fork of Babylon, and it's hitting an error on that syntax.

### Your Environment

| software         | version(s)
| ---------------- | -------
| Babel            |  master
| Babylon          | master (from the last week or so)
| node             | 9.2.0
| npm              | yarn 1.3.2
| Operating System | macOS 10.13.2

**Comment by Kovensky** [comment]
That... shouldn't work, `export default` is a value; `interface` is not a value. It's not allowed by typescript's own spec either:

https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#11342-export-default-declarations

This is either a bug in typescript or on its spec.

**Comment by alangpierce** [comment]
It was explicitly added in https://github.com/Microsoft/TypeScript/pull/16040 and is part of a list of new `export default` syntaxes that are in progress: https://github.com/Microsoft/TypeScript/issues/3792#issuecomment-303526468 , so my guess is the spec hasn't been updated.

**Comment by Kovensky** [comment]
I personally would prefer if they didn't do that and kept `export default` strictly with its value semantics, but rolling it back would be a breaking change... 😢

**Comment by simonbuchan** [comment]
@Kovensky It's implied by the fact that Typescript shares the same export namespace between types and values. It occasionally causes issues (in particular importing only types from a module with side-effects: what *should* it do?), but most of the time it's simple to pick up what's happening. Except when you can't do a thing with a type that you can with values, like here, where it is just confusing.

@Kovensky, @alangpierce Yeah, the spec is old, hasn't been touched since TS 1.8, with some touchup on 2016-01-20. See [all the open Spec issues](https://github.com/Microsoft/TypeScript/labels/Spec) - it looks like there's now some movement there, but this isn't mentioned.

**Comment by Kovensky** [comment]
It's just that `export default` specifically implies an evaluation of its right side, while other `export` forms don't.

But I guess TS can just say that `interface` are a `HoistableDeclaration`.

**Comment by simonbuchan** [comment]
> It's just that export default specifically implies an evaluation of its right side, while other export forms don't.

(I apologize in advance for my terrible habit of replying with far too much detail. It's hard to break!)

Not quite sure what this is saying, since in JS declarations are evaluated, but I'm going to assume you mean it implies an expression? Otherwise you are excluding all type exports :)

From my understanding, essentially there are four kinds of export in plain JS:

```js
const foo = 123, bar = "abc";

// 1. export name list
export { foo, bar as baz };

// 2. export declaration
export function f() {}
export class C {}
export const x = 123;
// equivalent to
function f() {}
class C {}
const x = 123;
export { f, C, x };

// 3. export default expression
export default 123;
// equivalent to
const _new_name = 123;
export { _new_name as default };

// 4. export default declaration
export default function g() {}
// or
export default class D() {}
// equivalent to
function g() {}
export { g as default };
// or
class D() {}
export { D as default };
```

Note that 2, 3 and 4 are all equivalent to 1.

Typescript originally added:

```ts
// 5. export type names in name lists
type T = number;
interface U { x: number; }
export { T, U as V }

// 6. export type declarations
export type W = number;
export interface X { x: number };
```

So 5 is the type version of 1, making type and value names comparable, and 6 is the type version of 2, making type and value declarations comparable. A type version of 3 runs into syntax issues:

```ts
// 7. can't export type expression
export default { x: number }; // oops, this is a value
```

But a type version of 4 is pretty straightforward, and is more surprising for being missing than not:

```ts
// 8. export default type declaration
export default interface Y { x: number };
// equivalent to:
interface Y { x: number };
export { Y as default };
```

(only this supported at the moment, other forms, such as `export default type T = ...;` [look to be on track](https://github.com/Microsoft/TypeScript/issues/3792#issuecomment-303526468))

If you are still unconvinced, for a motivating case, consider a standard practice of naming your files after their default export:

```js
// Storage.js
export default class Storage { ... }
```

Now say you split this into two implementations:

```js
// LocalStorage.js
import Storage from "./Storage";
export default class LocalStorage extends Storage { ... }
// RemoteStorage.js
import Storage from "./Storage";
export default class RemoteStorage extends Storage { ... }
```

Now you heard that types are what's cool right now so you rename all your files from .js to .ts. Later you realize that you don't need a `Storage` *class*, an interface would do fine. Naively:

```ts
// Storage.ts
export default interface Storage {}
```

Without `export default interface`, the alternatives are confusing, misleading or just ugly!

**Comment by Kovensky** [comment]
Exactly, `export default` implies an expression _except_ on the specific exceptions added for `HoistableDeclaration` and `ClassDeclaration`. Those exceptions just makes it so that those specific things that would have looked like assignment expressions (`FunctionExpression` / `ClassExpression`) are considered declarations instead.

An `interface` is not a value nor looks like a valid assignment expression so I don't see it working with `export default`, unless you could argue that it is a `HoistableDeclaration`. `export default type X` is definitely out because that'd be a `VariableDeclaration`, which is not acceptable.

It does turn out that `export default (expression that evaluates to a type)` also works too so things are already crazy 😖

**Comment by simonbuchan** [comment]
> those specific things that would have looked like assignment expressions

Not sure what this means? That `export default function foo() {}` should not be interpreted as `const _new_name = function foo() {}, export { _new_name as default }`? That's the same issue as for non-default export declarations, which TS already established a symmetry with type declarations `export interface foo {}`.

If it helps, here's a table:

JS | TS
---|---
`export const x = 123` | `export type T = number`
`export { x }` | `export { T }`
`export class Foo {}` | `export interface Foo {}`
`export default class Foo {}` | `export default interface Foo {}` *new*

**Comment by Kovensky** [comment]
Exactly: `export default function foo () {}` is _not_ the same as `const _default = function foo () {}; export { _default as default }`.

`export default` only allows `AssignmentExpression` (i.e. expressions that have a value that can be assigned to an lvalue). The `HoistableDeclaration` and `ClassDeclaration` exceptions exist to avoid turning those into expressions and keep their declaration nature.

`export default function foo () {}` is the same as `export { foo as default }; function foo () {}`. `export default function () {}` is also valid, and exports a hoisted function whose `name` is `*default*` (with the asterisks).

It's very important that it remains treated as a hoisted function so the behavior is still correct on circular dependencies, even if by not giving it a local name you cannot refer to it inside the module.

**Comment by simonbuchan** [comment]
I was asking if that was what you were referring to, not it's implications - I had already spec-ed that case's behavior above as export form 4 🙂

Your second paragraph seems weird: you say `export default` only allows expressions, then that it has declaration exceptions to preserve the declaration semantics? This is strange logic, expressions and declarations behave completely differently! It is slightly unfortunate that `function foo() {}` and `class Foo {}` are either expressions or declarations (with the resultant differences in semantics) depending on location, but it's hardly the worst language ambiguity in JS.

Surely it's clearer to say "export default allows either expressions or (some) declarations"? Where the excluded declarations in JS are, to my knowledge, only `var`/`let`/`const`, which are excluded because they can introduce multiple names, and it's not worth introducing another form just for export default. That is, it has two syntactic and semantic forms, which I've labeled 3 and 4 above.

By this interpretation, Typescript already had a type equivalent to forms 1 and 2, (forms 5 and 6) and they are adding the missing type equivalent to form 4, form 8.

In other words, I'm claiming that JS specs `Hoistable` and `ClassDeclaration` essentially as "all declarations except variables, cause we don't care about fixing it for `export default var a, b`", so when typescript adds a bunch of other declarations they naturally should be included.

(Incidentally, I'm not sure about `export default function () {}` - `function () {}` is an expression and not a declaration, clearly, and it behaves essentially the same in form 3 barring the function name. It could simply be getting that as a side-effect of transpilation, but if it's explicitly spec-ed as exporting a function named `default`, then you could simply treat it as another special syntax form and semantics for export default if it bothers you)

Sorry if I'm getting repetitive!

**Comment by Kovensky** [comment]
`export default` _only_ allows expressions except for things that fall under the lookahead `[lookahead ∉ { function, async [no LineTerminator here] function, class }]`, which is there specifically to allow `export default HoistableDeclaration` and `export default ClassDeclaration` to match. https://tc39.github.io/ecma262/#prod-ExportDeclaration

It's only `export` that allows arbitrary `Declaration`s (and `VariableStatement`).

This is by far what people confuse the most about `export default`, because `export default` is very special. `export default` is not an export named default, `export default` is not the default export, `export default` is an expression evaluation that is assigned to a concrete value named `*default*`, that is then exported as the export named `default`. It has two specific exceptions used to keep `function` and `class` behaving as declarations, instead of being treated as expressions. But `export default` _is a concrete value_.

`export default` is also the only exporter that has non-trivial runtime semantics (except for the `HoistableDeclaration` case).

**Comment by simonbuchan** [comment]
And `Declaration` is `HoistableDeclaration` or `ClassDeclaration` or `LexicalDeclaration`; that is "`HoistableDeclaration` or `ClassDeclaration`" is identical to "`Declaration` - `LexicalDeclaration`". The way in which the grammar is specified in the spec is so that it is unambiguous, and BNF languages don't generally provide features like "try A, otherwise try B", or "C, unless it's a D". Don't take the specific form of the grammar as a statement from TC39 that they only like functions and classes!

Your third statement reads just flat out wrong: `export default` is absolutely syntax for an export named `default`. See the equivalences above, 2 and 4 are identical except for the exported binding name being renamed to `default`, and precluding variables for syntactic reasons, and 3 is just 4 with an implicit variable being rebound as `default`. If you support `export { foo as bar }` correctly, you can implement all the others in terms of it.

If it helps: try "there is no `export <Expression>;` form, because what's it's name?", then maybe it's clearer that exporting declarations is the default (no pun intended), and it's `export default <Expression>` that's a special case (in that it creates a value without introducing a local name).

In any case, it's a *fait accompli*, [and more are coming](https://github.com/Microsoft/TypeScript/pull/18628), so I guess... get used to it?

**Comment by Bessonov** [comment]
run in the same issue after switch from at-loader to babel.

**Comment by loganfsmyth** [comment]
@Kovensky How would it sound to you to have this parse, but have the TS transform throw an exception saying it is unsupported in Babel? It seems like we should support it in the parser at least, since plenty of tooling would want to use it for static analysis, even if they don't use it for transformation.

**Comment by simonbuchan** [comment]
That would be weird. The TS transform otherwise is tracking new typescript features, the only transform exceptions are ones that require whole-program transformation (`const enum`, `namespace` [though I'm not convinced by that one (but thats a different story!)]). Like all pure-type decls, the transform is just to delete it.

Why would @Kovensky get to veto typescript language features just because he doesn't like them?

**Comment by Kovensky** [comment]
I'd prefer to leave this to someone in the actual TypeScript team (@andy-ms?).

`export default` is explicitly something that acts in the value space, while `interface` is something explicitly type only; using both in the same place doesn't make sense to me and people using it appear (to me) to misunderstand what `export default` is.

**Comment by loganfsmyth** [comment]
Sorry, I probably should have read more before commenting. I was thinking TS interfaces were something we already couldn't support, and this was just a case of parser failure, but that is not the case.

I'll be curious to head Andy's opinion. To me this seems like another example where TS's allowance of overlap in type-producing and value-producing bindings causes friction, the same way it does for `import` statements importing both types and values with way to statically tell which is which.

While I agree with @Kovensky that it's a frustrating that TS allows this, I also think we should support it since TS does. I think @simonbuchan's logic is sound. If we accept that standalone interface declarations are entirely deleted, I think it's reasonable to entirely delete `export default interface {}` too.

**Comment by simonbuchan** [comment]
> If we accept that standalone interface declarations are entirely deleted, I think it's reasonable to entirely delete export default interface {} too.

I was going to comment that `interface Foo {} export default Foo;` is already legal (though bad) TS, [but it looks like babel doesn't correctly elide the `export default` there](https://babeljs.io/repl/#?babili=false&browsers=&build=&builtIns=false&code_lz=JYOwLgpgTgZghgYwgAgGIHt3IN4CgC-uEAHgA7pRjIAmE8ArgDZUboDcuQA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&lineWrap=true&presets=typescript&prettier=false&targets=&version=7.0.0-beta.46&envVersion=) - but that's a different issue!

**Comment by eventualbuddha** [comment]
I'm looking into adding TypeScript support to babel-codemod and I ran into this issue because babel-codemod itself uses the `export default interface ...` pattern in exactly the way @simonbuchan says (see [Resolver.ts](https://github.com/square/babel-codemod/blob/9e44e2e0d47e2bcefc0f8de475ae85057e4649cf/src/resolvers/Resolver.ts)).

For what it's worth, I agree that babel should just delete the `export default interface` statement altogether when transforming TypeScript. I might have some time to work on this if nobody else is already going to do so.

**Comment by devdoomari3** [comment]
@eventualbuddha are you going to submit a PR? (plz do)

btw, @Kovensky reading your comments, I don't think the discussion led to anything except noise. Discussions about typescript syntax right/wrong should happen on a typescript github-issues page. 

Please don't surprise any user by differing from what the current TS compiler would do. Users don't care about whether that TS syntax is right/wrong, is standard/not.

PS: come to think about it... wouldn't using typescript transpiler API solve 99% of the issue related to ```@babel/plugin-transform-typescript``` ?

Or is there something more to babel-ecosystem?
